File: benchmarks/ocaml/nqueens.ml
type solution =
| Nil
| Cons of int * solution
;;

let rec safe queen diag xs =
  match xs with
  | Nil -> true
  | Cons (q, tail) ->
    queen <> q &&
    queen <> q + diag &&
    queen <> q - diag &&
    safe queen (diag + 1) tail
;;

let rec place n row soln =
  if row = 0 then 1
  else try_col n row soln n

and try_col n row soln col =
  if col <= 0 then 0
  else if safe col 1 soln then place n (row - 1) (Cons (col, soln)) + try_col n row soln (col - 1)
  else try_col n row soln (col - 1)
;;

let queens n =
  place n n Nil
;;

let () =
  Printf.printf "%d\n" (queens 13)

File: benchmarks/rust/nqueens.rs
#[derive(Clone)]
enum Solution {
    Nil,
    Cons(i64, Box<Solution>),
}

fn safe(queen: i64, diag: i64, xs: &Solution) -> bool {
    match xs {
        Solution::Nil => true,
        Solution::Cons(q, tail) => {
            queen != *q && queen != *q + diag && queen != *q - diag && safe(queen, diag + 1, tail)
        }
    }
}

fn place(n: i64, row: i64, soln: &Solution) -> i64 {
    if row == 0 {
        1
    } else {
        try_col(n, row, soln, n)
    }
}

fn try_col(n: i64, row: i64, soln: &Solution, col: i64) -> i64 {
    if col <= 0 {
        0
    } else if safe(col, 1, soln) {
        place(n, row - 1, &Solution::Cons(col, Box::new(soln.clone())))
            + try_col(n, row, soln, col - 1)
    } else {
        try_col(n, row, soln, col - 1)
    }
}

fn queens(n: i64) -> i64 {
    place(n, n, &Solution::Nil)
}

fn main() {
    println!("{}", queens(13));
}
